giftless/storage/aws_s3.py

```diff
@@ -0,0 +1,110 @@
+import os
+from typing import Any, BinaryIO, Dict, Iterable, Optional
+
+import boto3  # type: ignore
+import botocore  # type: ignore
+
+from giftless.storage import ExternalStorage, StreamingStorage
+from giftless.storage.exc import ObjectNotFound
+
+
+class AwsS3Storage(StreamingStorage, ExternalStorage):
+    """AWS S3 Blob Storage backend.
+    """
+
+    def __init__(self, aws_s3_bucket_name: str, path_prefix: Optional[str] = None, **_):
+        self.aws_s3_bucket_name = aws_s3_bucket_name
+        self.path_prefix = path_prefix
+        self.s3 = boto3.resource('s3')
+        self.s3_client = boto3.client('s3')
+
+    def get(self, prefix: str, oid: str) -> Iterable[bytes]:
+        if not self.exists(prefix, oid):
+            raise ObjectNotFound()
+        result: Iterable[bytes] = self._s3_object(prefix, oid).get()['Body']
+        return result
+
+    def put(self, prefix: str, oid: str, data_stream: BinaryIO) -> int:
+        completed = []
+
+        def upload_callback(size):
+            completed.append(size)
+
+        bucket = self.s3.Bucket(self.aws_s3_bucket_name)
+        bucket.upload_fileobj(data_stream, self._get_blob_path(prefix, oid), Callback=upload_callback)
+        return sum(completed)
+
+    def exists(self, prefix: str, oid: str) -> bool:
+        s3_object = self._s3_object(prefix, oid)
+        try:
+            s3_object.load()
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                return False
+            else:
+                raise RuntimeError(e)
+        return True
+
+    def get_size(self, prefix: str, oid: str) -> int:
+        if self.exists(prefix, oid):
+            result: int = self._s3_object(prefix, oid).content_length
+            return result
+        else:
+            raise ObjectNotFound()
+
+    def get_upload_action(self, prefix: str, oid: str, size: int, expires_in: int,
+                          extra: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
+        params_ = {
```

Comment:
why the extra `_` at the end of the variable name? Do we have another object in the scope with the same name?

Reply:
I have used the _ to indicate that it's only a tmp variable which exists only for sake of readability.
I have removed the trailing `_` not to create confusion.